controllers/stardoginstance_controller.go

```diff
@@ -7,36 +7,293 @@ package controllers
+	}
+)
+
+const stardogUsersFinalizer = "finalizer.stardog.users"
+const stardogRolesFinalizer = "finalizer.stardog.roles"
 
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardoginstances,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardoginstances/status,verbs=get;update;patch
-
 func (r *StardogInstanceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("stardoginstance", req.NamespacedName)
+	namespace := req.NamespacedName
+	_ = r.Log.WithValues("stardoginstance", namespace)
+
+    ctx := context.Background()
+    stardogInstance := &StardogInstance{}
+
+    err := r.Client.Get(ctx, namespace, stardogInstance)
+    if err != nil {
+        if apierrors.IsNotFound(err) {
+            r.Log.Info("StardogInstance not found, ignoring reconcile.", "StardogInstance", namespace)
+            return ctrl.Result{Requeue: false}, nil
+        }
+        r.Log.Error(err, "Could not retrieve StardogInstance.", "StardogInstance", namespace)
+        return ctrl.Result{Requeue: true, RequeueAfter: r.ReconcileInterval}, err
+    }
+
+	rc := &ReconciliationContext{
+		ctx:        ctx,
+		cfg:        stardogInstance,
+		conditions: make(map[StardogConditionType]StardogCondition),
+	}
+
+	r.Log.Info("reconciling", getLoggingKeysAndValuesForStardogInstance(stardogInstance)...)
+
+	//TODO find out how to delete when credentials do not work or the StardogInstance never connected to a Stardog DB
+	isStardogInstanceMarkedToBeDeleted := stardogInstance.GetDeletionTimestamp() != nil
+	if isStardogInstanceMarkedToBeDeleted {
+		if r.deleteStardogInstance(rc) != nil {
+			rc.SetStatusCondition(CreateStatusConditionToBeDeleted(err))
+			rc.SetStatusCondition(CreateStatusConditionReady(false))
+			return ctrl.Result{Requeue: true}, r.updateStatus(rc)
+		}
+	}
+
+	if r.validateSpecification(rc.cfg.Spec) != nil {
+		rc.SetStatusCondition(CreateStatusConditionInvalid(err))
+		rc.SetStatusCondition(CreateStatusConditionReady(false))
+		return ctrl.Result{}, r.updateStatus(rc)
+	}
+	rc.SetStatusIfExisting(StardogInvalid, v1.ConditionFalse)
+
+	if r.validateConnection(rc) != nil {
+		rc.SetStatusCondition(CreateStatusConditionErrored(err))
+		rc.SetStatusCondition(CreateStatusConditionReady(false))
+		return ctrl.Result{}, r.updateStatus(rc)
+	}
+
+	if missingAtLeastOne(rc.cfg.GetFinalizers(), stardogUsersFinalizer, stardogRolesFinalizer) {
+		if r.addDeletionFinalizers(rc) != nil {
+			rc.SetStatusCondition(CreateStatusConditionErrored(err))
+			rc.SetStatusCondition(CreateStatusConditionReady(false))
+			return ctrl.Result{Requeue: true}, r.updateStatus(rc)
+		}
+	}
+
+	rc.SetStatusCondition(CreateStatusConditionReady(true))
+
+	return ctrl.Result{}, r.updateStatus(rc)
+}
+
+func (r *StardogInstanceReconciler) addDeletionFinalizers(rc *ReconciliationContext, finalizers ...string) error {
+	r.Log.Info("adding Finalizers for the StardogInstance")
```

Comment:
That's debug information
```suggestion
	r.Log.V(1).Info("adding Finalizers for the StardogInstance")
```